File: src/main/java/net/vidageek/games/regex/Tasks.java
package net.vidageek.games.regex;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import net.vidageek.games.task.Task;

public class Tasks {

	private List<Task> tasks = new ArrayList<Task>(); 
	
	public void add(final Task task) {
		this.tasks.add(task.withIndex(tasks.size()));
	}

	public Task in(final int index) {
		return this.tasks.get(index);
	}

	public int size() {
		return tasks.size();
	}

	public List<Task> all() {
		return Collections.unmodifiableList(this.tasks);
	}

}

File: src/main/scala/vggames/shared/task/Tasks.scala
package vggames.shared.task

import scala.collection.JavaConverters.seqAsJavaListConverter
import scala.collection.mutable.ListBuffer

class Tasks(taskGroups : TaskGroup*) {

  val tasks : List[IndexedTask[_]] = unmodifiableIndexedTasks

  def at(index : Int) = tasks(index)

  def size = tasks.size

  def all : java.util.List[IndexedTask[_]] = tasks.asJava

  private def unmodifiableIndexedTasks = {
    taskGroups.foldLeft((0, List[IndexedTask[_]]())) { (acc, group) =>
      group.foldLeft(acc) { (acc, task) =>
        (acc._1 + 1, acc._2 :+ new IndexedTask(new GroupedTask(group, task), acc._1))
      }
    }._2
  }
}